### ISSUE TITLE
Unable to provide Repository with immutable Credentials

### ISSUE THREAD

**Issue by kalleand** [ajoberstar/grgit#163]
When upgrading from 1.7.0 to 1.7.1 of gradle-git it is no longer possible to provide Credentials to a Repository like this
```
def grgit = org.ajoberstar.grgit.Grgit.open(creds: new org.ajoberstar.grgit.Credentials(<username>, <password>))
```

This now fails with
```
> @Immutable processor doesn't know how to handle field 'credentials' of type 'org.ajoberstar.grgit.Credentials' while constructing class org.ajoberstar.grgit.Repository.
  @Immutable classes only support properties with effectively immutable types including:
  - Strings, primitive types, wrapper types, Class, BigInteger and BigDecimal, enums
  - other @Immutable classes and known immutables (java.awt.Color, java.net.URI)
  - Cloneable classes, collections, maps and arrays, and other classes with special handling (java.util.Date)
  Other restrictions apply, please see the groovydoc for @Immutable for further details
```

My very brief attempt at troubleshooting this issue leads to this commit: 13eb6c9a14830e74309cb19fd9a88ff889765f73 where Credentials is changed from `@Immutable` to `@Canonical` while Repository is still annotated with `@Immutable`.

**Comment by ajoberstar** [comment]
Thanks for catching this. It will be in the next bug fix release.

**Comment by sugun999** [comment]
Hi,

Was this issue fixed?

**Comment by ajoberstar** [comment]
Hoping to have some time today. The alternative is to go back to an older release.

**Comment by ajoberstar** [comment]
1.9.3-rc.2 should fix this issue. If someone can verify, I'll make a final 1.9.3 and update gradle-git's dependency as well.

**Comment by ajoberstar** [comment]
Just note that's 1.9.3-rc.2 of grgit, you would still use 1.7.1 of gradle-git (for now).

**Comment by borkur** [comment]
I had an issue with building the 1.9.3-rc.2  
`oracle@bigdatalite grgit-1.9.3-rc.2]$ ./gradlew --info
Initialized native services in: /home/oracle/.gradle/native
Connected to daemon DaemonInfo{pid=29647, address=[55b5914a-7892-4be3-a65f-2a2c730416b2 port:12192, addresses:[/0:0:0:0:0:0:0:1, /127.0.0.1]], state=Idle, lastBusy=1496925746356, context=DefaultDaemonContext[uid=e05673a4-2722-41ba-88ab-655f5ff59293,javaHome=/usr/java/jdk1.8.0_111,daemonRegistryDir=/home/oracle/.gradle/daemon,pid=29647,idleTimeout=10800000,daemonOpts=-XX:MaxPermSize=256m,-XX:+HeapDumpOnOutOfMemoryError,-Xmx1024m,-Dfile.encoding=UTF-8,-Duser.country=US,-Duser.language=en,-Duser.variant]}. Dispatching request Build{id=f8a0f951-1e0a-4043-85f0-ec672e9641b2.1, currentDir=/home/oracle/GIT/grgit-1.9.3-rc.2}.
Received result org.gradle.launcher.daemon.protocol.BuildStarted@37e4d7bb from daemon DaemonInfo{pid=29647, address=[55b5914a-7892-4be3-a65f-2a2c730416b2 port:12192, addresses:[/0:0:0:0:0:0:0:1, /127.0.0.1]], state=Idle, lastBusy=1496925746356, context=DefaultDaemonContext[uid=e05673a4-2722-41ba-88ab-655f5ff59293,javaHome=/usr/java/jdk1.8.0_111,daemonRegistryDir=/home/oracle/.gradle/daemon,pid=29647,idleTimeout=10800000,daemonOpts=-XX:MaxPermSize=256m,-XX:+HeapDumpOnOutOfMemoryError,-Xmx1024m,-Dfile.encoding=UTF-8,-Duser.country=US,-Duser.language=en,-Duser.variant]} (build should be starting).
The client will now receive all logging from the daemon (pid: 29647). The daemon log file: /home/oracle/.gradle/daemon/3.3/daemon-29647.out.log
Starting 11th build in daemon [uptime: 3 hrs 4 mins 53.24 secs, performance: 100%, no major garbage collections]
Executing build with daemon context: DefaultDaemonContext[uid=e05673a4-2722-41ba-88ab-655f5ff59293,javaHome=/usr/java/jdk1.8.0_111,daemonRegistryDir=/home/oracle/.gradle/daemon,pid=29647,idleTimeout=10800000,daemonOpts=-XX:MaxPermSize=256m,-XX:+HeapDumpOnOutOfMemoryError,-Xmx1024m,-Dfile.encoding=UTF-8,-Duser.country=US,-Duser.language=en,-Duser.variant]
Creating new cache for plugin-use-metadata, path /home/oracle/.gradle/caches/3.3/plugin-resolution/plugin-use-metadata.bin, access org.gradle.cache.internal.DefaultCacheAccess@1d863f83
Creating new cache for client-status, path /home/oracle/.gradle/caches/3.3/plugin-resolution/client-status.bin, access org.gradle.cache.internal.DefaultCacheAccess@1d863f83
Starting Build
Settings evaluated using settings file '/home/oracle/GIT/grgit-1.9.3-rc.2/settings.gradle'.
Projects loaded. Root project using build file '/home/oracle/GIT/grgit-1.9.3-rc.2/build.gradle'.
Included projects: [root project 'grgit-1.9.3-rc.2']
Evaluating root project 'grgit-1.9.3-rc.2' using build file '/home/oracle/GIT/grgit-1.9.3-rc.2/build.gradle'.
Creating new cache for metadata-2.23/module-metadata, path /home/oracle/.gradle/caches/modules-2/metadata-2.23/module-metadata.bin, access org.gradle.cache.internal.DefaultCacheAccess@3d7f8595
Creating new cache for metadata-2.23/artifact-at-repository, path /home/oracle/.gradle/caches/modules-2/metadata-2.23/artifact-at-repository.bin, access org.gradle.cache.internal.DefaultCacheAccess@3d7f8595
No git repository found. Build may fail with NPE.

FAILURE: Build failed with an exception.

* Where:
Build file '/home/oracle/GIT/grgit/grgit/build.gradle' line: 2

* What went wrong:
An exception occurred applying plugin request [id: 'org.ajoberstar.defaults', version: '0.7.3']
> Failed to apply plugin [id 'org.ajoberstar.git-publish']
   > Could not get unknown property 'grgit' for root project 'grgit-1.9.3-rc.2' of type org.gradle.api.Project.

* Try:
Run with --stacktrace option to get the stack trace. Run with --debug option to get more log output.

BUILD FAILED

Total time: 0.963 secs
Received result Failure[value=org.gradle.initialization.ReportedException: org.gradle.internal.exceptions.LocationAwareException: Build file '/home/oracle/GIT/grgit/grgit/build.gradle' line: 2
An exception occurred applying plugin request [id: 'org.ajoberstar.defaults', version: '0.7.3']] from daemon DaemonInfo{pid=29647, address=[55b5914a-7892-4be3-a65f-2a2c730416b2 port:12192, addresses:[/0:0:0:0:0:0:0:1, /127.0.0.1]], state=Idle, lastBusy=1496925746356, context=DefaultDaemonContext[uid=e05673a4-2722-41ba-88ab-655f5ff59293,javaHome=/usr/java/jdk1.8.0_111,daemonRegistryDir=/home/oracle/.gradle/daemon,pid=29647,idleTimeout=10800000,daemonOpts=-XX:MaxPermSize=256m,-XX:+HeapDumpOnOutOfMemoryError,-Xmx1024m,-Dfile.encoding=UTF-8,-Duser.country=US,-Duser.language=en,-Duser.variant]} (build should be done).`

However, when I copied the src directory over to the 1.10.0 and did a build, it worked - and the #163 issue seems to be resolved.

**Comment by ajoberstar** [comment]
Thanks for confirming the fix! 1.9.3 final is now out. 

The error you encountered is a lingering issue with running builds using grgit run without a Git repo.
